aten/src/ATen/native/cuda/AmpKernels.cu

```diff
@@ -0,0 +1,93 @@
+#define _USE_MATH_DEFINES
+
+#include <math.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cuda/Loops.cuh>
+
+
+namespace at {
+namespace native {
+
+// These functions have no CPU variants, so there's no need to bother with a .cpp file in native/, and
+// also no need to impose the DECLARE/DEFINE/REGISTER_DISPATCH layer of indirection.
+
+Tensor & _amp_unscale_inf_check_cuda(Tensor & scaled_grad,
+                                     const Tensor & current_scale,
+                                     const Tensor & found_inf)
+{
+  // Do I really need all these checks?
+  TORCH_CHECK(scaled_grad.is_cuda(), "scaled_grad must be a CUDA tensor.");
+  TORCH_CHECK(current_scale.is_cuda(), "current_scale must be a CUDA tensor.");
+  TORCH_CHECK(found_inf.is_cuda(), "found_inf must be a CUDA tensor.");
+  TORCH_CHECK(current_scale.numel() == 1, "current_scale must be a 1-element tensor.");
+  TORCH_CHECK(found_inf.numel() == 1, "found_inf must be a 1-element tensor.");
+  TORCH_CHECK(current_scale.scalar_type() == at::ScalarType::Float, "current_scale must be a float tensor.");
+  TORCH_CHECK(found_inf.scalar_type() == at::ScalarType::Float, "found_inf must be a float tensor.");
+
+  auto rscale = current_scale.to(at::ScalarType::Double).reciprocal();
+
+  // Act on scaled_grad in place.
+  auto iter = TensorIterator::unary_op(scaled_grad, scaled_grad);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    iter.dtype(),
+    "_amp_unscale_inf_check_kernel",
+    [&] {
+      float* found_inf_ptr = found_inf.data_ptr<float>();
+      double* rscale_ptr = rscale.data_ptr<double>();
+
+      gpu_kernel(iter, [found_inf_ptr, rscale_ptr] GPU_LAMBDA(scalar_t val) -> scalar_t {
+          float fval = static_cast<float>(val);
+          if(!std::isfinite(fval))
+            *found_inf_ptr = 1.f;
+          return static_cast<scalar_t>(fval*(*rscale_ptr));
+        });
+    });
+
+  return scaled_grad;
+}
+
+
+// Update the scale factor Tensor on the device.
+__global__ void amp_update_scale_kernel(float* current_scale,
+                                        float* found_inf,
+                                        float* new_scale,
+                                        double scale_growth_factor,
+                                        double scale_backoff_factor)
+{
+  // This kernel should only ever be launched with one thread anyway, but just in case
```

Comment:
You control how this function is called. I don't think you need this conditional?

Reply:
Answering this question for Carilli,
We do the conditional check inside the kernel, so we can keep the loss_scale_factor on the device instead of copying it back to host to make that decision on host code.

This is necessary to remove the host/device sync at the end of each iteration.

Reply:
Thanks Jie :). 
I think @mruberry was referring to the threadIdx.x and blockIdx.x check, not the inf state check.  The threadIdx.x and blockIdx.x check can be gotten rid of.